Orange/widgets/data/owsave.py

```diff
@@ -34,170 +19,270 @@ class OWSave(widget.OWWidget):
+    def save_file_as(self):
+        filename, selected_filter = self.get_save_filename()
         if not filename:
             return
-
         self.filename = filename
+        self.filter = selected_filter
         self.last_dir = os.path.split(self.filename)[0]
-        self.basename = os.path.basename(self.remove_extensions(filename))
-        self.unconditional_save_file()
-        self.adjust_label()
+        self.bt_save.setText(f"Save as {os.path.split(filename)[1]}")
+        self._update_messages()
+        self.save_file()
 
-    def save_file(self):
+    def _update_messages(self):
+        self.Error.no_file_name(
+            shown=not self.filename and self.auto_save)
+        self.Error.unsupported_sparse(
+            shown=self.data is not None and self.data.is_sparse()
+            and self.filename and not self.writer.SUPPORT_SPARSE_DATA)
+        self.Warning.type_annotation_ignored(
+            shown=self.data is not None and self.filename
+            and self.add_type_annotations
+            and not self.writer.OPTIONAL_TYPE_ANNOTATIONS)
+
+    def _update_status(self):
         if self.data is None:
-            return
-        if not self.filename:
-            self.save_file_as()
+            self.info.set_input_summary(self.info.NoInput)
+        else:
+            self.info.set_input_summary(
+                str(len(self.data)),
+                f"Data set {self.data.name or '(no name)'} "
+                f"with {len(self.data)} instances")
+
+    def send_report(self):
+        self.report_data_brief(self.data)
+        writer = self.writer
+        noyes = ["No", "Yes"]
+        self.report_items((
+            ("File name", self.filename or "not set"),
+            ("Format", writer.DESCRIPTION),
+            ("Type annotations",
+             writer.OPTIONAL_TYPE_ANNOTATIONS
+             and noyes[self.add_type_annotations])
+        ))
+
+    @classmethod
+    def migrate_settings(cls, settings, version=0):
+        def migrate_to_version_2():
+            # Set the default; change later if possible
+            settings.pop("compression", None)
+            settings["filter"] = next(iter(cls.filters))
+            filetype = settings.pop("filetype", None)
+            if filetype is None:
+                return
+
+            ext = cls._extension_from_filter(filetype)
+            if settings.pop("compress", False):
+                for afilter in cls.filters:
+                    if ext + ".gz" in afilter:
+                        settings["filter"] = afilter
+                        return
+                # If not found, uncompressed may have been erroneously set
+                # for a writer that didn't support if (such as .xlsx), so
+                # fall through to uncompressed
+            for afilter in cls.filters:
+                if ext in afilter:
+                    settings["filter"] = afilter
+                    return
+
+        if version < 2:
+            migrate_to_version_2()
```

Comment:
Migrations are in a local function because the code uses `return`-s to abort migration if it fails or when it's done. Not using `return`-s would make the code overly `if`-i -- this is way simpler. However, if these were returns from `migrate_setting`, they would complicate future migrations. Hence a local function.